xla/service/gpu/transforms/cudnn_custom_call_compiler.cc

```diff
@@ -541,47 +246,134 @@ absl::StatusOr<se::gpu::CudnnGraph> HloCustomCallToCuDnnGraph(
+        bmm1_grad_gemm2_rhs_type, bmm1_grad_gemm2_rhs_shape.dimensions(),
+        bmm1_grad_gemm2_rhs_shape.layout().minor_to_major(),
+        bmm1_grad_gemm2_dnums.rhs_batch_dimensions(),
+        bmm1_grad_gemm2_dnums.rhs_contracting_dimensions());
+
+    MatmulTensorDescriptor bmm2_grad_gemm1_lhs = MatmulTensorDescriptor::For(
+        bmm2_grad_gemm1_lhs_type, bmm2_grad_gemm1_lhs_shape.dimensions(),
+        bmm2_grad_gemm1_lhs_shape.layout().minor_to_major(),
+        bmm2_grad_gemm1_dnums.lhs_batch_dimensions(),
+        bmm2_grad_gemm1_dnums.lhs_contracting_dimensions());
+
+    MatmulTensorDescriptor d_output = MatmulTensorDescriptor::For(
+        d_output_type, d_output_shape.dimensions(),
+        d_output_shape.layout().minor_to_major(),
+        bmm2_grad_gemm1_dnums.rhs_batch_dimensions(),
+        bmm2_grad_gemm1_dnums.rhs_contracting_dimensions());
+
+    MatmulTensorDescriptor bmm2_grad_gemm2_rhs = MatmulTensorDescriptor::For(
+        bmm2_grad_gemm2_rhs_type, bmm2_grad_gemm2_rhs_shape.dimensions(),
+        bmm2_grad_gemm2_rhs_shape.layout().minor_to_major(),
+        bmm2_grad_gemm2_dnums.rhs_batch_dimensions(),
+        bmm2_grad_gemm2_dnums
+            .rhs_contracting_dimensions());  // FMHA TODO: transpose here?
+
+    TensorDescriptor d_bmm1_lhs =
+        TensorDescriptor::For(d_bmm1_lhs_type, d_bmm1_lhs_shape.dimensions(),
+                              d_bmm1_lhs_shape.layout().minor_to_major());
+    TensorDescriptor d_bmm1_rhs =
+        TensorDescriptor::For(d_bmm1_rhs_type, d_bmm1_rhs_shape.dimensions(),
+                              d_bmm1_rhs_shape.layout().minor_to_major());
+    TensorDescriptor d_bmm2_rhs =
+        TensorDescriptor::For(d_bmm2_rhs_type, d_bmm2_rhs_shape.dimensions(),
+                              d_bmm2_rhs_shape.layout().minor_to_major());
+
+    std::optional<se::dnn::TensorDescriptor> d_bias;
+    if (d_bias_shape.has_value()) {
+      // Get DNN dtype from primtive types
+      TF_ASSIGN_OR_RETURN(
+          DataType d_bias_type,
+          GetDNNDataTypeFromPrimitiveType(d_bias_shape->element_type()));
+      d_bias = TensorDescriptor::For(d_bias_type, d_bias_shape->dimensions(),
+                                     d_bias_shape->layout().minor_to_major());
+    }
+
+    std::optional<se::dnn::TensorDescriptor> mask;
+    if (mask_shape.has_value()) {
+      TF_ASSIGN_OR_RETURN(DataType mask_type, GetDNNDataTypeFromPrimitiveType(
+                                                  mask_shape->element_type()));
+      mask = TensorDescriptor::For(mask_type, mask_shape->dimensions(),
+                                   mask_shape->layout().minor_to_major());
+    }
+
+    std::optional<se::dnn::TensorDescriptor> fwd_output;
+    if (fwd_output_shape.has_value()) {
+      TF_ASSIGN_OR_RETURN(
+          DataType fwd_output_type,
+          GetDNNDataTypeFromPrimitiveType(fwd_output_shape->element_type()));
+      fwd_output =
```

Comment:
Did you forget to use all these optionals or am I missing something?

Reply:
This means that they were not in use already before I started simplifying. Removed a few more.